app/src/org/commcare/dalvik/odk/provider/InstanceProvider.java

```diff
@@ -365,10 +387,140 @@ public int update(Uri uri, ContentValues values, String where, String[] whereArg
                 throw new IllegalArgumentException("Unknown URI " + uri);
         }
 
+        // If we've changed a particular form instance's status, and not
+        // created a new entry (hence count > 0 check), we need to mirror the
+        // change in that form's record.  NOTE: this conditional is crucial,
+        // since updating a form record in turn calls this update function, so
+        // we need to break the infinite loop by only updating the form record
+        // when the the status changes.
+        if (values.containsKey(InstanceColumns.STATUS) && count > 0) {
+            try {
+                linkToSessionFormRecord(getInstanceRowUri(uri, where, whereArgs));
+            } catch (Exception e) {
+                throw new SQLException("Failed to update row " + uri);
+            }
+        }
+
         getContext().getContentResolver().notifyChange(uri, null);
+
         return count;
     }
 
+    /**
+     * Check if a URI points to a concrete instance; if it doesn't
+     * then rebuild the uri from the result of a query using the where
+     * arguments.
+     *
+     * @param potentialUri URI pointing to the instance table or a particular
+     * entry in that table
+     * @param selection A selection criteria to apply when filtering rows. If
+     * null then all rows are included.
+     * @param selectionArgs You may include ?s in selection, which will be
+     * replaced by the values from selectionArgs, in order that they appear in
+     * the selection. The values will be bound as Strings.
+     * @return URI pointing to a row in the instance table, either the one passed
+     * in or built from a query using the method arguments.
+     */
+    private Uri getInstanceRowUri(Uri potentialUri, String selection, String[] selectionArgs) {
```

Comment:
Newly added to make sure we only try to link a form record to a uri pointing to a concrete instance.